pkg/cloud/service_test.go

```diff
@@ -22,63 +22,104 @@ import (
-	fo := &fakeOperation{isDoneFunc: func(ctx context.Context) (bool, error) {
-		attempts++
-		if attempts < totalAttempts {
-			return false, nil
-		}
-		return true, nil
-	}}
-	s := Service{RateLimiter: &NopRateLimiter{}}
-	// Check that pollOperation will retry the operation multiple times.
-	err := s.pollOperation(context.Background(), fo)
-	if err != nil {
-		t.Errorf("pollOperation() = %v, want nil", err)
-	}
-	if attempts != totalAttempts {
-		t.Errorf("`attempts` = %d, want %d", attempts, totalAttempts)
-	}
+/*
+Copyright 2018 Google LLC
 
-	// Check that the operation's error is returned.
-	fo.err = fmt.Errorf("test operation failed")
-	err = s.pollOperation(context.Background(), fo)
-	if err != fo.err {
-		t.Errorf("pollOperation() = %v, want %v", err, fo.err)
-	}
-	fo.err = nil
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
 
-	fo.isDoneFunc = func(ctx context.Context) (bool, error) {
-		return false, nil
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloud
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+)
+
+func TestPollOperation(t *testing.T) {
+	testErr := errors.New("test error")
+	tests := []struct {
+		name                  string
+		op                    *fakeOperation
+		cancel                bool
+		wantErr               error
+		wantRemainingAttempts int
+	}{
+		{
+			name: "Retry",
+			op:   &fakeOperation{attemptsRemaining: 10},
+		},
+		{
+			name: "OperationFailed",
+			op: &fakeOperation{
+				attemptsRemaining: 2,
+				err:               testErr,
+			},
+			wantErr: testErr,
+		},
+		{
+			name: "DoneFailed",
+			op: &fakeOperation{
+				attemptsRemaining: 2,
+				doneErr:           testErr,
+			},
+			wantErr: testErr,
+		},
+		{
+			name:                  "Cancel",
+			op:                    &fakeOperation{attemptsRemaining: 1},
+			cancel:                true,
+			wantErr:               context.Canceled,
+			wantRemainingAttempts: 1,
+		},
 	}
-	// Use context that has been cancelled and expect a context error returned.
-	ctxCancelled, cancelled := context.WithCancel(context.Background())
-	cancelled()
-	// Verify context is cancelled by now.
-	<-ctxCancelled.Done()
-	// Check that pollOperation returns because the context is cancelled.
-	err = s.pollOperation(ctxCancelled, fo)
-	if err == nil {
-		t.Errorf("pollOperation() = nil, want: %v", ctxCancelled.Err())
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			s := Service{RateLimiter: &NopRateLimiter{}}
+			ctx, cfn := context.WithTimeout(context.Background(), 3*time.Second)
```

Comment:
we should use 1ms or less (I'm assuming test does not rely on actual timeout)

Reply:
The timeout is to have decent failures if the test deadlocks (ie. without the corresponding change in service.go). 1ms looks extremely low, can you really bet there will be enough CPU in testing envs to guarantee everything completes within 1ms? 3s seems much more practical - and it kicks in only if the code deadlocks anyway.

Reply:
As long as we aren't having unit tests waiting around for 3 s each, sgtm